lib/akasha/storage/memory_event_store/stream.rb

```diff
@@ -3,13 +3,18 @@ module Storage
     class MemoryEventStore
       # Memory-based event stream.
       class Stream
-        def initialize
+        # Creates a new event stream.
+        # Accepts an optional block, with reduce function used in `write_events`
+        # to accumulate new events. By default, it just appends new events to the
+        # stream but it can be used for filtering, etc.
+        def initialize(&reduce)
+          @reduce = reduce || ->(all_events, new_events) { all_events + new_events }
```

Comment:
Hum. This interface seems a bit... constraining? I know it looks awesome and simple, but it forces us to work on Arrays, instead of lazy maps - which could provide punishing in case of large streams, no?
In the spirit of Ruby, we could make `all_events` an instance of our custom class which implements lazy append and `Enum` methods, but that's not so great. What's the purpose of this `reduce` anyway? If projections, then we need only simple `map` / `flat_map`, not `reduce` ;)

Reply:
1. This is in-memory stream used for tests, not likely to have many events.
2. I like flatmap too but it's constraining on other fronts. I actually refactored it into `reduce`. 

?

Reply:
But you'll want to keep the same interface for other event stores, no?

Reply:
No. Right now interface is only necessary between `MemoryStore` and `MemoryStore::Stream` to emulate projections and is not a public one, so it can easily be changed.

I see the merit of establishing an interface that will work for, say, Postgres but the interface doesn't have to be the same.

Still, I modified it, hope you don't terribly dislike it. :> I don't want to force a quasi-functional interface so went with more "idiomatic" Ruby (`before_xx`).